### PR TITLE
Add codelens action for running concrete playback tests

### LINKED ISSUE #52: Create unit test functionality for concrete playback
- [ ]  Add a `concrete_playback` button in the diff message generated when the proof returns as a Failure.
- [ ] Add a code lens action that runs the `concrete_playback` unit test using ` cargo test`.

### PR DESCRIPTION
**Pull request by jaisnan**
### Description of changes: 

Why - The regular buttons for Running tests and Debugging will not work as the functions used by concrete playback are out of scope (due to invoking harnesses which are inside the `cfg(kani)` scope).

This PR Adds the `codelens` button (`Run Tests (Kani)`) which enables users to run their concrete playback unit tests just like their other unit tests by emulating the flow for other unit tests but injecting kani's scope into the final command being called.

Screen Shot 2023-05-03 at 7 59 56 PM

The result of clicking on the button

Screen Shot 2023-05-03 at 8 01 43 PM
### Resolved issues:

Resolves #52 

### Call-outs:

1. Uses parallel parsing using the tree-sitter to add all of the codelens actions i.e there are two parallel parsing operations that happen on the source code, that could be sweeped into one for efficiency.

### Testing:

* How is this change tested? Unit tests

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/ui/CodeLensProvider.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	 * @param _token - Unused token
+	 * @returns
+	 */
+
+	public provideCodeLenses(
+		document: vscode.TextDocument,
+		_token: vscode.CancellationToken,
+	): vscode.CodeLens[] | Thenable<vscode.CodeLens[]> {
+		if (vscode.workspace.getConfiguration('codelens-sample').get('enableCodeLens', true)) {
+			this.codeLenses = [];
```
[the remark this answers is not in the document]
Comment (jaisnan) [reply src/ui/CodeLensProvider.ts:37]: `this.codelens `is a class variable for the class we're extending so I couldn't rename it unfortunately.

File: src/model/runCargoTest.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+export async function runCodeLensTest(functionName: string): Promise<void> {
+	const taskName = `Cargo Test: ${functionName}`;
+
+	const packageName = await getPackageName();
+	const taskDefinition: vscode.TaskDefinition = {
+		type: 'cargo',
+		command: 'test',
+		args: [functionName],
+	};
+	const cargoTestCommand: string = `RUSTFLAGS="--cfg=kani" cargo test --package ${packageName} --bin ${packageName} -- ${functionName} --exact --nocapture`;
```
[the remark this answers is not in the document]
Comment (jaisnan) [reply src/model/runCargoTest.ts:20]: `--bin ${packageName}` prevents recompiling when there's no changes to the unit tests and the cargo test runner recommends it in it's warnings so I included it. We could exclude it though if needed.

File: src/test/test-programs/concretePlaybackTests.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+export const rustFileWithUnitTestsOnly = `
+#[cfg(test)]
+mod test {
+    #[test]
+    #[cfg_attr(miri, ignore)] // this test is too expensive for miri
```
[the remark this answers is not in the document]
Comment (jaisnan) [reply src/test/test-programs/concretePlaybackTests.ts]: Removed them. Thanks for pointing it out.

File: src/extension.ts [hunk 4]
```diff
@@ -235,3 +256,12 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
 	context.subscriptions.push(runningViewerReport);
 	context.subscriptions.push(runningConcretePlayback);
 }
+
+// this method is called when your extension is deactivated
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function deactivate() {
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply src/extension.ts:265]: This isn't needed, it could have its own PR. It's called when the extension is disabled.

File: src/extension.ts [hunk 5]
```diff
@@ -13,13 +14,16 @@ import {
...
 } from './test-tree/createTests';
+import { CodelensProvider } from './ui/CodeLensProvider';
 import { callConcretePlayback } from './ui/concrete-playback/concretePlayback';
 import { callViewerReport } from './ui/reportView/callReport';
 import { showInformationMessage } from './ui/showMessage';
 import { SourceCodeParser } from './ui/sourceCodeParser';
 import { startWatchingWorkspace } from './ui/watchWorkspace';
 import { checkCargoExist, getContentFromFilesystem, getRootDirURI } from './utils';
 
+let disposables: vscode.Disposable[] = [];
```
Comment (adpaco-aws) [inline src/extension.ts:25]: Remove this too.

File: src/extension.ts [hunk 6]
```diff
@@ -224,6 +228,26 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
 		data.addToCrate(controller, file, treeRoot);
 	}
 
+	const codelensProvider = new CodelensProvider();
+	const rustLanguageSelector = { scheme: 'file', language: 'rust' };
+
+	vscode.languages.registerCodeLensProvider(rustLanguageSelector, codelensProvider);
+
+	// Allows VSCode to enable VSCode globally (if the user switches off code lens in settings, the kani codelens action will be too)
```
Comment (adpaco-aws) [inline src/extension.ts:236]:
```suggestion
	// Allows VSCode to enable code lens globally.
	// If the user switches off code lens in settings, the Kani code lens action will be switched off too.
```

File: src/extension.ts [hunk 7]
```diff
@@ -224,6 +228,23 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
...
 	}
 
+	const codelensProvider = new CodelensProvider();
+	const rustLanguageSelector = { scheme: 'file', language: 'rust' };
+
+	vscode.languages.registerCodeLensProvider(rustLanguageSelector, codelensProvider);
+
+	vscode.commands.registerCommand('codelens-sample.enableCodeLens', () => {
+		vscode.workspace.getConfiguration('codelens-sample').update('enableCodeLens', true, true);
+	});
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply src/extension.ts:239]: Solved in discussion.

File: src/model/runCargoTest.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply src/model/runCargoTest.ts:20]: Add that as a comment.

File: src/ui/CodeLensProvider.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+				// This is the metadata that VSCode needs to place the codelens button
+				const line = document.lineAt(startPosition.row);
+				const indexOf = line.text.indexOf(function_item_name);
+				const position = new vscode.Position(line.lineNumber, indexOf);
+				const range = document.getWordRangeAtPosition(position, new RegExp(this.regex));
+
+				const runTestAction = {
+					title: 'Run Test (Kani)',
+					tooltip: 'Run unit test generated by kani',
```
Comment (adpaco-aws) [inline src/ui/CodeLensProvider.ts:60]:
```suggestion
					tooltip: 'Run unit test generated by Kani',
```

File: src/ui/sourceCodeParser.ts [hunk 9]
```diff
@@ -120,6 +120,55 @@ export namespace SourceCodeParser {
...
+				result.push([function_item_name, line]);
+			}
+		}
+
+		return result;
+	}
+
+	// Find all concrete playback generated unit tests using tree walking
+	export function findKaniTests(rootNode: any): any[] {
+		// Find all attributes with #[test], then filter those with the concrete_playback_prefix
```
Comment (adpaco-aws) [inline src/ui/sourceCodeParser.ts:150]:
```suggestion
		// Find all attributes with `#[test]`, then filter those with the `concrete_playback` prefix
```

File: src/model/runCargoTest.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	const taskName = `Cargo Test: ${functionName}`;
+
+	const packageName = await getPackageName();
+	const taskDefinition: vscode.TaskDefinition = {
+		type: 'cargo',
+		command: 'test',
+		args: [functionName],
+	};
+
+	// Adding --bin {packageName} to the command prevents recompiling when there's no changes to the unit tests, which makes rerunning unit tests faster.
```
Comment (adpaco-aws) [inline src/model/runCargoTest.ts:21]:
```suggestion
	// Adding `--bin {packageName}` to the command prevents recompiling when
	// there's no changes to the unit tests, which makes rerunning unit tests faster.
```

**Review by adpaco-aws — COMMENTED** [review]
Just one suggestion on a long comment, otherwise good to go. Thanks!

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**